include/oboe/AudioStreamBuilder.h

```diff
@@ -354,11 +354,31 @@ class AudioStreamBuilder : public AudioStreamBase {
      * <strong>Important: See AudioStreamCallback for restrictions on what may be called
      * from the callback methods.</strong>
      *
+     * We pass a shared_ptr so that the sharedDataCallback object cannot be deleted
+     * before the stream is deleted.
+     *
      * @param dataCallback
      * @return pointer to the builder so calls can be chained
      */
-    AudioStreamBuilder *setDataCallback(oboe::AudioStreamDataCallback *dataCallback) {
+    AudioStreamBuilder *setDataCallback(std::shared_ptr<AudioStreamDataCallback> sharedDataCallback) {
+        // Use this raw pointer in the rest of the code to retain backwards compatibility.
+        mDataCallback = sharedDataCallback.get();
```

Comment:
Are we sure that this works as intended? There is nothing stopping the raw pointer from getting deleted somewhere else in the code.

Reply:
We control Oboe and can promise not to delete the raw pointer. Maybe we should add a comment to that effect. We did not create the callback and should not directly delete it. That would be a bad bug.

The callback object is created by the app and shared with Oboe. Oboe AudioStream will hold a shared pointer so we know the callback will not get deleted by the app while the AudioStream still exists. And if the AudioStream is created with a shared_ptr then we know it will exist for the duration of the error callback.

It is always possible to delete some raw pointer and mess things up if we try. The goal is to make it possible for apps to  be safe if they try. Before this change, it was hard for apps to be safe even when they tried.